Use WorkerThreadPool for Server threads (enhanced)

RandomShaper opened the pull request:
This PR is the version of #77004 where my collaboration on top of @reduz's work has been added.

This is kept from the description of the original PR:
> * Servers now use WorkerThreadPool for background computation.
> * This helps keep the number of threads used fixed at all times.
> * It also ensures everything works on HTML5 with threads.
> * And makes it easier to support disabling threads for also HTML5.

This is new:
> `CommandQueueMT` now syncs with the servers via the `WorkerThreadPool` yielding mechanism, which makes its classic main sync semaphore superfluous.
> Also, some warnings about calls that kill performance when using threaded rendering are removed because there's a mechanism that warns about that in a more general fashion.

This is no longer true:
> ~The only downside of this is that GLES3 threads likely no longer work because every time the render thread is called, the thread context is different. This needs to be researched how to fix. Maybe if GLES3 is used, threaded rendering server should be disabled for the time being?~

Supersedes #77004.

**2024-04-09:** A way of easily testing this being sure all the physics and rendering are threaded is to apply this patch:
[all_servers_mt.txt](https://github.com/godotengine/godot/files/14920984/all_servers_mt.txt)

reduz committed:
Use WorkerThreadPool for Server threads

* Servers now use WorkerThreadPool for background computation.
* This helps keep the number of threads used fixed at all times.
* It also ensures everything works on HTML5 with threads.
* And makes it easier to support disabling threads for also HTML5.

CommandQueueMT now syncs with the servers via the WorkerThreadPool
yielding mechanism, which makes its classic main sync semaphore
superfluous.

Also, some warnings about calls that kill performance when using
threaded rendering are removed because there's a mechanism that
warns about that in a more general fashion.

Co-authored-by: Pedro J. Estébanez <pedrojrulez@gmail.com>

AThousandShips commented on core/config/engine.cpp:
```diff
@@ -82,6 +82,16 @@ int Engine::get_audio_output_latency() const {
+	} else {
+		server_syncs = 0;
+	}
+	frame_server_synced = 0;
```
```suggestion
	frame_server_synced = false;
```

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -389,83 +391,115 @@ Error WorkerThreadPool::wait_for_task_completion(TaskID p_task_id) {
+}
+
+void WorkerThreadPool::notify_yield_over(TaskID p_task_id) {
+	task_mutex.lock();
```
```suggestion
	MutexLock lock(task_mutex);
```
Safer

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -389,83 +391,115 @@ Error WorkerThreadPool::wait_for_task_completion(TaskID p_task_id) {
+	task_mutex.lock();
+	Task **taskp = tasks.getptr(p_task_id);
+	if (!taskp) {
 		task_mutex.unlock();
```
```suggestion
```

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -389,83 +391,115 @@ Error WorkerThreadPool::wait_for_task_completion(TaskID p_task_id) {
+	td.signaled = true;
+	td.cond_var.notify_one();
+
+	task_mutex.unlock();
```
```suggestion
```

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -389,83 +391,115 @@ Error WorkerThreadPool::wait_for_task_completion(TaskID p_task_id) {
+		}
+	} else {
+		task->done_semaphore.wait();
+		task_mutex.lock();
```
```suggestion
		MutexLock lock(task_mutex);
```

RandomShaper replied:
Regarding the RAII vs. explicit style of locks, judging by code in other lowish level areas of the engine, I induce there's an unwritten rule to prefer explicit. That's why I've been continuist and kept using that style. There's a case of `MutexLock` below, but that's needed for the `ConditionVariable`. I can live with any of both styles, but explicit lock-unlock looks to me like the closest we have to a rule so far.

AThousandShips commented on core/object/worker_thread_pool.cpp:
```diff
@@ -389,83 +391,115 @@ Error WorkerThreadPool::wait_for_task_completion(TaskID p_task_id) {
+			tasks.erase(p_task_id);
+			task_allocator.free(task);
+		}
+		task_mutex.unlock();
```
```suggestion
```

bruvzg commented on platform/macos/display_server_macos.mm:
```diff
@@ -3056,9 +3056,25 @@
 }
 
 void DisplayServerMacOS::release_rendering_thread() {
+#if defined(GLES3_ENABLED)
+	if (gl_manager_angle) {
+		gl_manager_angle->release_current();
+	}
+	if (gl_manager_legacy) {
+		gl_manager_legacy->release_current();
+	}
+#endif
 }
 
 void DisplayServerMacOS::make_rendering_thread() {
+#if defined(GLES3_ENABLED)
+	if (gl_manager_angle) {
+		gl_manager_angle->make_current();
+	}
+	if (gl_manager_legacy) {
+		gl_manager_legacy->make_current();
+	}
+#endif
```
This seems to be called from the thread when it is started and set the current GL context (to the context of the active window), but GL context is already set from `_draw_viewport()` at the start of each window drawing, so it's likely redundant (probably part of the old 3.x single windows code).

RandomShaper replied:
That seems to be the case indeed. However, I'm not clear on what the right API would be because we now have the concepts of windows and threads somewhat conflated. I'm wondering if the right call would be the following:
- Remove `DisplayServer::make`/`release_rendering_thread()` .
- Remove `make`/release`_current()` from the various GL context classes.
- Modify `window_make_current()` so that taking `INVALID_WINDOW_ID` means releasing the context.

(When I say _remove_, it may be needed to be understood as _deprecate until we can remove_.)

RandomShaper replied:
In the next push, I'm replacing the commit where I added those by one that simplifies the APIs (without breaking compat, I think). `DisplayServer` as well as the various GL managers can therefore now have a window-specific make current, a global release, and call it a day.

bruvzg replied:
For the reference:

Multithreaded compatibility renderer was crashing (in the `GLES3::Utilities::get_video_adapter_api_version()`) before this PR. With it, it's no longer crashing, but still displaying only black screen. Multithreaded Vulkan renderer also not rendering anything (both with and without this PR), might be macOS specific issue, I'll check it later.

bruvzg replied:
In both cases, it deadlocks inside `blit_render_targets_to_screen`.

```
Main thread:

   🔻 Waiting for CommandQueueMT.
std::__1::recursive_mutex::lock() (@std::__1::recursive_mutex::lock():7)
CommandQueueMT::Command1<RendererCanvasCull, void (RendererCanvasCull::*)(RID), RID>* CommandQueueMT::allocate_and_lock<CommandQueueMT::Command1<RendererCanvasCull, void (RendererCanvasCull::*)(RID), RID>>() (core/templates/command_queue_mt.h:362)
void CommandQueueMT::push<RendererCanvasCull, void (RendererCanvasCull::*)(RID), RID>(RendererCanvasCull*, void (RendererCanvasCull::*)(RID), RID) (core/templates/command_queue_mt.h:407)
RenderingServerDefault::canvas_item_clear(RID) (servers/rendering/rendering_server_default.h:916)
CanvasItem::_redraw_callback() (scene/main/canvas_item.cpp:137)
CallQueue::_call_function(Callable const&, Variant const*, int, bool) (core/object/message_queue.cpp:221)
CallQueue::flush() (core/object/message_queue.cpp:326)
SceneTree::process(double) (scene/main/scene_tree.cpp:537)
    🔻 Inside main evnet loop.
Main::iteration() (main/main.cpp:4013)
OS_MacOS::run() (platform/macos/os_macos.mm:778)
main (platform/macos/godot_main_macos.mm:84)
start (@start:593)
```


```
Rendering thread:

    🔻 Since nothing seems directy holding it in the Engine code, I assue it's waiting for the next main event loop iteration, but main thread is locked.
_dlock_wait (@_dlock_wait:17)
_dispatch_thread_event_wait_slow (@_dispatch_thread_event_wait_slow:17)
-[CAMetalLayer(MoltenVK) screenMVK] (@-[CAMetalLayer(MoltenVK) screenMVK]:31)
MVKSwapchain::initSurfaceImages(VkSwapchainCreateInfoKHR const*, unsigned int) (@MVKSwapchain::initSurfaceImages(VkSwapchainCreateInfoKHR const*, unsigned int):153)
MVKSwapchain::MVKSwapchain(MVKDevice*, VkSwapchainCreateInfoKHR const*) (@MVKSwapchain::MVKSwapchain(MVKDevice*, VkSwapchainCreateInfoKHR const*):183)
MVKDevice::createSwapchain(VkSwapchainCreateInfoKHR const*, VkAllocationCallbacks const*) (@MVKDevice::createSwapchain(VkSwapchainCreateInfoKHR const*, VkAllocationCallbacks const*):15)
vkCreateSwapchainKHR (@vkCreateSwapchainKHR:31)
RenderingDeviceDriverVulkan::swap_chain_resize(RenderingDeviceDriver::CommandQueueID, RenderingDeviceDriver::SwapChainID, unsigned int) (drivers/vulkan/rendering_device_driver_vulkan.cpp:2663)
RenderingDevice::screen_prepare_for_drawing(int) (servers/rendering/rendering_device.cpp:3190)
RendererCompositorRD::blit_render_targets_to_screen(int, BlitToScreen const*, int) (servers/rendering/renderer_rd/renderer_compositor_rd.cpp:37)
RendererViewport::draw_viewports(bool) (servers/rendering/renderer_viewport.cpp:861)
RenderingServerDefault::_draw(bool, double) (servers/rendering/rendering_server_default.cpp:89)
    🔻 This lock CommandQueueMT mutex.
CommandQueueMT::_flush() (core/templates/command_queue_mt.h:377)
CommandQueueMT::flush_all() (core/templates/command_queue_mt.h:423)
RenderingServerDefault::_thread_loop() (servers/rendering/rendering_server_default.cpp:347)
...
```

bruvzg replied:
Probably not a solution, but MT renderer works with the following changes, until something in the main thread will call `RenderingServer` in a wrong time:

```diff
diff --git a/core/templates/command_queue_mt.h b/core/templates/command_queue_mt.h
index 71030d86d1..76b25d75e4 100644
--- a/core/templates/command_queue_mt.h
+++ b/core/templates/command_queue_mt.h
@@ -366,12 +366,17 @@ class CommandQueueMT {
 
 	void _flush() {
 		lock();
+		uint64_t flush_read_ptr_tmp = flush_read_ptr;
+		LocalVector<uint8_t> command_mem_tmp = command_mem;
+		command_mem.clear();
+		flush_read_ptr = 0;
+		unlock();
 
 		WorkerThreadPool::thread_enter_command_queue_mt_flush(this);
-		while (flush_read_ptr < command_mem.size()) {
-			uint64_t size = *(uint64_t *)&command_mem[flush_read_ptr];
-			flush_read_ptr += 8;
-			CommandBase *cmd = reinterpret_cast<CommandBase *>(&command_mem[flush_read_ptr]);
+		while (flush_read_ptr_tmp < command_mem_tmp.size()) {
+			uint64_t size = *(uint64_t *)&command_mem_tmp[flush_read_ptr_tmp];
+			flush_read_ptr_tmp += 8;
+			CommandBase *cmd = reinterpret_cast<CommandBase *>(&command_mem_tmp[flush_read_ptr_tmp]);
 
 			SyncSemaphore *sync_sem = cmd->get_sync_semaphore();
 			cmd->call();
@@ -379,20 +384,16 @@ class CommandQueueMT {
 				sync_sem->sem.post(); // Release in case it needs sync/ret.
 			}
 
-			if (unlikely(flush_read_ptr == 0)) {
+			if (unlikely(flush_read_ptr_tmp == 0)) {
 				// A reentrant call flushed.
-				DEV_ASSERT(command_mem.is_empty());
+				DEV_ASSERT(command_mem_tmp.is_empty());
 				unlock();
 				return;
 			}
 
-			flush_read_ptr += size;
+			flush_read_ptr_tmp += size;
 		}
 		WorkerThreadPool::thread_exit_command_queue_mt_flush();
-
-		command_mem.clear();
-		flush_read_ptr = 0;
-		unlock();
 	}
 
 	void wait_for_flush();
diff --git a/platform/macos/display_server_macos.mm b/platform/macos/display_server_macos.mm
index 854ee59ed8..42cf3b1ce8 100644
--- a/platform/macos/display_server_macos.mm
+++ b/platform/macos/display_server_macos.mm
@@ -2158,7 +2158,7 @@ void DisplayServerMacOS::window_set_size(const Size2i p_size, WindowID p_window)
 }
 
 Size2i DisplayServerMacOS::window_get_size(WindowID p_window) const {
-	_THREAD_SAFE_METHOD_
+	//_THREAD_SAFE_METHOD_
 
 	ERR_FAIL_COND_V(!windows.has(p_window), Size2i());
 	const WindowData &wd = windows[p_window];
diff --git a/platform/macos/godot_content_view.mm b/platform/macos/godot_content_view.mm
index 93bba84783..9d76fded26 100644
--- a/platform/macos/godot_content_view.mm
+++ b/platform/macos/godot_content_view.mm
@@ -55,9 +55,11 @@
 - (void)displayLayer:(CALayer *)layer {
 	DisplayServerMacOS *ds = (DisplayServerMacOS *)DisplayServer::get_singleton();
 	if (OS::get_singleton()->get_main_loop() && ds->get_is_resizing() && need_redraw) {
-		Main::force_redraw();
-		if (!Main::is_iterating()) { // Avoid cyclic loop.
-			Main::iteration();
+		if ((OS::get_singleton()->get_render_thread_mode() != OS::RENDER_SEPARATE_THREAD)) {
+			Main::force_redraw();
+			if (!Main::is_iterating()) { // Avoid cyclic loop.
+				Main::iteration();
+			}
 		}
 		need_redraw = false;
 	}
```

There seems to be a lot of various conditions multithreaded renderer might lock up, it's also constantly crashing in with out of bounds access in the texture storage. Since there's a "WARNING: The Multi-Threaded rendering thread model is experimental. However, bear in mind that at the moment it can lead to project crashes or instability." warning, I'll assume it is expected to be completely broken right now.

bruvzg replied:
Checked it with 4.2.1, Vulkan MT renderer seems to be working, OpenGL is crashing, I'll try bisecting to see what broke it later.

RandomShaper replied:
I only really tested on Windows. I had to change how locking was working in DisplayServer to prevent deadlocks. I tried to propagate the changes across the other platforms, but maybe I didn't it well. I may test on Mac myself to see what's going on.

bruvzg replied:
Bisected it to https://github.com/godotengine/godot/pull/87340 - Finish splitting functionality of the `RenderingDevice` backends into `RenderingDeviceDriver` (it's big, so probably won't be easy to detect what exactly is causing the issue).

RandomShaper replied:
@bruvzg I realized a missing unlock in my changes to MacOS. Would you please re-test? If it deadlocks, please test with #90470 as well.

bruvzg replied:
Neither is working.

Since it was originally broken by https://github.com/godotengine/godot/pull/87340, which is not changing `CommandQueueMT` in any way, it's probably not directly related, and the issue is in `RenderingDevice`.

In some cases (resizing window for example), it still lockups with the aforementioned patch (and lockups only involve `RenderingDevice` not `CommandQueueMT`).

RandomShaper replied:
Alright. Thanks for your help! I'll delve.

RandomShaper replied:
The more I look into this, the more issues with multi-threading rendering (physics seem OK) I'm seeing. My suggestion (CC @akien-mga) is that we merge this as it is, given it seems it doesn't break single-threading, and I'll make a subsequent PR fixing MT rendering.

RandomShaper commented on doc/classes/WorkerThreadPool.xml:
```diff
@@ -109,5 +116,15 @@
+		<method name="yield">
+			<return type="void" />
+			<description>
+				Called from a task running on a worker thread, lets such thread run other tasks.
+				This method has to be called whenever the task has nothing to do and can therefore enter a collaborative wait state.
+				Another thread will call [method notify_yield_over] when the task should stop yielding the thread and resume its operation.
+				This is a way of implementing a daemon-like task, which is long-lived, in a way that it doesn't monopolize a worker thread.
```
It seems I thought I was posting on X, or something. 😃

RandomShaper committed:
WorkerThreadPool: Fix data race

RandomShaper committed:
Polish interaction between windowing, input and rendering

- Adapt GL make/release API to the current architecture.
- Fix DisplayServer being locked while dispatching input (prevent deadlocks).

RandomShaper committed:
WorkerThreadPool: Support daemon-like tasks (via yield semantics)

AThousandShips reviewed (approved):
Can't vouch for the thread pool code but the rest looks clean to me

akien-mga commented:
Tested against the GDQuest TPS demo with multithreaded rendering and physics. Rendering seems to work ok, but physics gave me a crash. It worked fine initially but after breaking a few crates things got laggy and eventually segfaulted.

```
ERROR: FATAL: Index p_index = 0 is out of bounds (shapes.size() = 0).
   at: get_shape_transform (./servers/physics_3d/godot_collision_object_3d.h:128)
ERROR: Caller thread can't call this function in this node (/root). Use call_deferred() or call_thread_group() instead.
   at: propagate_notification (./scene/main/node.cpp:2420)

================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.3.dev.custom_build (67b5e1779070c61ac0595301ec1c5080e1765b05)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib64/libc.so.6(+0x3e9a0) [0x7fa3307889a0] (??:0)
[2] GodotCollisionObject3D::get_shape_transform(int) const (/home/akien/Godot/godot/./servers/physics_3d/godot_collision_object_3d.h:128 (discriminator 3))
[3] GodotArea2Pair3D::setup(float) (/home/akien/Godot/godot/./servers/physics_3d/godot_area_pair_3d.cpp:126 (discriminator 3))
[4] GodotStep3D::_setup_constraint(unsigned int, void*) (/home/akien/Godot/godot/./servers/physics_3d/godot_step_3d.cpp:118)
[5] WorkerThreadPool::GroupUserData<GodotStep3D, void (GodotStep3D::*)(unsigned int, void*), decltype(nullptr)>::callback_indexed(unsigned int) (/home/akien/Godot/godot/./core/object/worker_thread_pool.h:185)
[6] WorkerThreadPool::_process_task(WorkerThreadPool::Task*) (/home/akien/Godot/godot/./core/object/worker_thread_pool.cpp:92)
[7] WorkerThreadPool::_thread_function(void*) (/home/akien/Godot/godot/./core/object/worker_thread_pool.cpp:195)
[8] Thread::callback(unsigned long, Thread::Settings const&, void (*)(void*), void*) (/home/akien/Godot/godot/./core/os/thread.cpp:66)
[9] void std::__invoke_impl<void, void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*>(std::__invoke_other, void (*&&)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long&&, Thread::Settings&&, void (*&&)(void*), void*&&) (/usr/include/c++/13/bits/invoke.h:61)
[10] std::__invoke_result<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*>::type std::__invoke<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*>(void (*&&)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long&&, Thread::Settings&&, void (*&&)(void*), void*&&) (/usr/include/c++/13/bits/invoke.h:97)
[11] void std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul>) (/usr/include/c++/13/bits/std_thread.h:292)
[12] std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> >::operator()() (/usr/include/c++/13/bits/std_thread.h:299)
[13] std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(unsigned long, Thread::Settings const&, void (*)(void*), void*), unsigned long, Thread::Settings, void (*)(void*), void*> > >::_M_run() (/usr/include/c++/13/bits/std_thread.h:244)
[14] /home/akien/Godot/godot/bin/godot.linuxbsd.editor.dev.x86_64() [0x7f882f3] (thread.o:?)
[15] /lib64/libc.so.6(+0x8e897) [0x7fa3307d8897] (??:0)
[16] /lib64/libc.so.6(+0x115a5c) [0x7fa33085fa5c] (??:0)
-- END OF BACKTRACE --
================================================================
```

RandomShaper commented:
Tryin to reproduce, to no avail, so far. Maybe it's one of those obscure and unpredictable platform-specific things. I'm wondering, could you try reproducing on `master` (I mean, with threaded physics)?

akien-mga commented:
> Trying to reproduce, to no avail, so far. Maybe it's one of those obscure and unpredictable platform-specific things. I'm wondering, could you try reproducing on `master` (I mean, with threaded physics)?

It indeed seems to be reproducible on `master` with multi-threaded physics, with the same backtrace. So it's not caused by this PR. I'll open a separate bug report.

---

So I guess we can merge this PR now and get wider testing.

akien-mga commented:
Thanks! Great work :tada:
